File: android/src/main/java/com/supermap/rnsupermap/JSCollectorElement.java
package com.supermap.rnsupermap;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.supermap.RNUtils.JsonUtil;
import com.supermap.data.Geometry;
import com.supermap.data.GeometryType;
import com.supermap.data.Point2D;
import com.supermap.data.Point2Ds;
import com.supermap.data.Rectangle2D;
import com.supermap.mapping.collector.CollectorElement;
import com.supermap.mapping.collector.CollectorElement;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by Yangshanglong on 2018/7/20.
 */
public class JSCollectorElement extends ReactContextBaseJavaModule {
    public static final String REACT_CLASS = "JSCollectorElement";
    private static Map<String,CollectorElement> mCollectorElementList = new HashMap();
    CollectorElement mCollectorElement;

    public JSCollectorElement(ReactApplicationContext context){super(context);}

    public static CollectorElement getObjFromList(String id) { return mCollectorElementList.get(id); }

    public static String registerId(CollectorElement collector){
        if(!mCollectorElementList.isEmpty()) {
            for(Map.Entry entry:mCollectorElementList.entrySet()){
                if(collector.equals(entry.getValue())){
                    return (String)entry.getKey();
                }
            }
        }

        Calendar calendar=Calendar.getInstance();
        String id=Long.toString(calendar.getTimeInMillis());
        mCollectorElementList.put(id,collector);
        return id;
    }

    @Override
    public String getName(){
        return REACT_CLASS;
    }

    /**
     * 添加点
     * @param collectorId
     * @param point2DId
     * @param promise
     */
    @ReactMethod
    public void addPoint(String collectorId, String point2DId, Promise promise){
        try{
            CollectorElement collector = mCollectorElementList.get(collectorId);
            Point2D point2D = JSPoint2D.getObjFromList(point2DId);
            collector.addPoint(point2D);
            promise.resolve(true);
        }catch(Exception e){
            promise.reject(e);
        }
    }

    /**
     * 通过 Geomotry 构造采集对象
     * @param collectorId
     * @param geometryId
     * @param promise
     */
    @ReactMethod
    public void fromGeometry(String collectorId, String geometryId, Promise promise){
        try{
            CollectorElement collector = mCollectorElementList.get(collectorId);
            Geometry geometry = JSGeometry.getObjFromList(geometryId);
            boolean result = collector.fromGeometry(geometry);
            promise.resolve(result);
        }catch(Exception e){
            promise.reject(e);
        }
    }

    /**
     * 获取采集对象的边框范围
     * @param collectorId
     * @param promise
     */
    @ReactMethod
    public void getBounds(String collectorId, Promise promise){
        try{
            CollectorElement collector = mCollectorElementList.get(collectorId);
            Rectangle2D rectangle2D = collector.getBounds();
            String rectangle2DId = JSRectangle2D.registerId(rectangle2D);
            promise.resolve(rectangle2DId);
        }catch(Exception e){
            promise.reject(e);
        }
    }

    /**
     * 获取采集对象的 Geometry
     * @param collectorId
     * @param promise
     */
    @ReactMethod
    public void getGeometry(String collectorId, Promise promise){
        try{
            CollectorElement collector = mCollectorElementList.get(collectorId);
            Geometry geometry = collector.getGeometry();
            String geometryId = JSGeometry.registerId(geometry);
            promise.resolve(geometryId);
        }catch(Exception e){
            promise.reject(e);
        }
    }

    /**
     * 获取采集对象的几何对象类型
     * @param collectorId
     * @param promise
     */
    @ReactMethod
    public void getGeometryType(String collectorId, Promise promise){
        try{
            CollectorElement collector = mCollectorElementList.get(collectorId);
            GeometryType type = collector.getGeometryType();
            int typeValue = type.value();
            promise.resolve(typeValue);
        }catch(Exception e){
            promise.reject(e);
        }
    }

    /**
     * 获取采集对象的点串
     * @param collectorId
     * @param promise
     */
    @ReactMethod
    public void getGeoPoints(String collectorId, Promise promise){
        try{
            CollectorElement collector = mCollectorElementList.get(collectorId);
            Point2Ds point2Ds = collector.getGeoPoints();
            WritableArray array = JsonUtil.point2DsToJson(point2Ds);

            promise.resolve(array);
        }catch(Exception e){
            promise.reject(e);
        }
    }

    /**
     * 获取采集对象的ID
     * @param collectorId
     * @param promise
     */
    @ReactMethod
    public void getID(String collectorId, Promise promise){
        try{
            CollectorElement collector = mCollectorElementList.get(collectorId);
            int value = collector.getID();

            promise.resolve(value);
        }catch(Exception e){
            promise.reject(e);
        }
    }

    /**
     * 获取采集对象的名称
     * @param collectorId
     * @param promise
     */
    @ReactMethod
    public void getName(String collectorId, Promise promise){
        try{
            CollectorElement collector = mCollectorElementList.get(collectorId);
            String value = collector.getName();

            promise.resolve(value);
        }catch(Exception e){
            promise.reject(e);
        }
    }

    /**
     * 获取采集对象的备注信息
     * 获取采集对象的名称
     * @param collectorId
     * @param promise
     */
    @ReactMethod
    public void getNotes(String collectorId, Promise promise){
        try{
            CollectorElement collector = mCollectorElementList.get(collectorId);
            String value = collector.getNotes();

            promise.resolve(value);
        }catch(Exception e){
            promise.reject(e);
        }
    }

    /**
     * 获取单击事件监听器
     * @param collectorId
     * @param promise
     */
    @ReactMethod
    public void getOnClickListenner(String collectorId, Promise promise){
        try{
            CollectorElement collector = mCollectorElementList.get(collectorId);
            CollectorElement.OnClickListener listener = collector.getOnClickListenner();

            // TODO 获取单击事件监听器
            promise.resolve(true);
        }catch(Exception e){
            promise.reject(e);
        }
    }

    /**
     * 获取点串分组信息（仅适用于通过Geomotry构造的动态数据）
     * @param collectorId
     * @param promise
     */
    @ReactMethod
    public void getPart(String collectorId, Promise promise){
        try{
            CollectorElement collector = mCollectorElementList.get(collectorId);
            ArrayList part = collector.getPart();
            WritableArray array = Arguments.fromArray(part);

            promise.resolve(array);
        }catch(Exception e){
            promise.reject(e);
        }
    }

    /**
     * 获取采集对象的类型
     * @param collectorId
     * @param promise
     */
    @ReactMethod
    public void getType(String collectorId, Promise promise){
        try{
            CollectorElement collector = mCollectorElementList.get(collectorId);
            CollectorElement.GPSElementType type = collector.getType();
            String typeName = type.name();

            promise.resolve(typeName);
        }catch(Exception e){
            promise.reject(e);
        }
    }

    /**
     * 获取用户数据
     * @param collectorId
     * @param promise
     */
    @ReactMethod
    public void getUserData(String collectorId, Promise promise){
        try{
            CollectorElement collector = mCollectorElementList.get(collectorId);
            Object object = collector.getUserData();

            promise.resolve(object);
        }catch(Exception e){
            promise.reject(e);
        }
    }

    /**
     * 设置采集对象的名称
     * @param collectorId
     * @param value
     * @param promise
     */
    @ReactMethod
    public void setName(String collectorId, String value, Promise promise){
        try{
            CollectorElement collector = mCollectorElementList.get(collectorId);
            collector.setName(value);

            promise.resolve(true);
        }catch(Exception e){
            promise.reject(e);
        }
    }

    /**
     * 设置采集对象的备注信息
     * @param collectorId
     * @param value
     * @param promise
     */
    @ReactMethod
    public void setNotes(String collectorId, String value, Promise promise){
        try{
            CollectorElement collector = mCollectorElementList.get(collectorId);
            collector.setNotes(value);

            promise.resolve(true);
        }catch(Exception e){
            promise.reject(e);
        }
    }

    /**
     * 设置点击监听器
     * @param collectorId
     * @param value
     * @param promise
     */
    @ReactMethod
    public void setOnClickListenner(String collectorId, String value, Promise promise){
        try{
            CollectorElement collector = mCollectorElementList.get(collectorId);
            // TODO 设置点击监听器

            promise.resolve(true);
        }catch(Exception e){
            promise.reject(e);
        }
    }

    /**
     * 设置用户数据
     * @param collectorId
     * @param value
     * @param promise
     */
    @ReactMethod
    public void setUserData(String collectorId, WritableMap value, Promise promise){
        try{
            CollectorElement collector = mCollectorElementList.get(collectorId);
            collector.setUserData(value);

            promise.resolve(true);
        }catch(Exception e){
            promise.reject(e);
        }
    }

}

File: ios/objNative/JSCollectorElement.m
//
//  JSCollectorElement.m
//  Supermap
//
//  Created by imobile-xzy on 2018/8/1.
//  Copyright © 2018年 Facebook. All rights reserved.
//

#import "JSCollectorElement.h"
#import "JSObjManager.h"
#import "SuperMap/CollectorElement.h"


@implementation JSCollectorElement
RCT_EXPORT_MODULE();

/**
 * 添加点
 * @param collectorId
 * @param point2DId
 * @param promise
 */
RCT_REMAP_METHOD(addPoint,addGPSPointById:(NSString*)senderId  point:(NSString*)point resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        CollectorElement* sender = [JSObjManager getObjWithKey:senderId];
        Point2D* point = [JSObjManager getObjWithKey:point];
        [sender addPoint:point];
        resolve([NSNumber numberWithBool:YES]);
        
    } @catch (NSException *exception) {
        reject(@"JSCollectorElement",@"addPoint expection",nil);
    }
}
/**
 * 通过 Geomotry 构造采集对象
 * @param collectorId
 * @param geometryId
 * @param promise
 */
RCT_REMAP_METHOD(fromGeometry,fromGeometry:(NSString*)senderId  geometry:(NSString*)geometryId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        CollectorElement* sender = [JSObjManager getObjWithKey:senderId];
        Geometry* geo = [JSObjManager getObjWithKey:geometryId];
        BOOL result = [sender fromGeometry:geo];
        resolve([NSNumber numberWithBool:result]);
        
    } @catch (NSException *exception) {
        reject(@"JSCollectorElement",@"fromGeometry expection",nil);
    }
}


/**
 * 获取采集对象的边框范围
 * @param collectorId
 * @param promise
 */
RCT_REMAP_METHOD(getBounds,getBounds:(NSString*)senderId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        CollectorElement* sender = [JSObjManager getObjWithKey:senderId];
        Rectangle2D* bounds = [sender getBounds];
        NSInteger nsKey = (NSInteger)bounds;
        [JSObjManager addObj:bounds];
        resolve(@(nsKey).stringValue);
    } @catch (NSException *exception) {
        reject(@"JSCollectorElement",@"getBounds expection",nil);
    }
}

/**
 * 获取采集对象的 Geometry
 * @param collectorId
 * @param promise
 */
RCT_REMAP_METHOD(getGeometry,getGeometry:(NSString*)senderId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        CollectorElement* sender = [JSObjManager getObjWithKey:senderId];
        Geometry *geo = [sender geometry];
        NSInteger nsKey = (NSInteger)geo;
        [JSObjManager addObj:geo];
        resolve(@(nsKey).stringValue);
    } @catch (NSException *exception) {
        reject(@"JSCollectorElement",@"getGeometry expection",nil);
    }
}

/**
 * 获取采集对象的几何对象类型
 * @param collectorId
 * @param promise
 */
RCT_REMAP_METHOD(getGeometryType,getGeometryType:(NSString*)senderId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        CollectorElement* sender = [JSObjManager getObjWithKey:senderId];
        GeometryType type = [sender getGeometryType];
        resolve([NSNumber numberWithInt:type]);
    } @catch (NSException *exception) {
        reject(@"JSCollectorElement",@"getGeometryType expection",nil);
    }
}

/**
 * 获取采集对象的点串
 * @param collectorId
 * @param promise
 */
RCT_REMAP_METHOD(getGeoPoints,getGeoPoints:(NSString*)senderId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        CollectorElement* sender = [JSObjManager getObjWithKey:senderId];
        Point2Ds *pnts = [sender getGeoPoints];
        NSInteger nsKey = (NSInteger)pnts;
        [JSObjManager addObj:pnts];
        resolve(@(nsKey).stringValue);
    } @catch (NSException *exception) {
        reject(@"JSCollectorElement",@"getGeoPoints expection",nil);
    }
}

/**
 * 获取采集对象的ID
 * @param collectorId
 * @param promise
 */
RCT_REMAP_METHOD(getID,getID:(NSString*)senderId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        CollectorElement* sender = [JSObjManager getObjWithKey:senderId];
        int idValue = [sender getID];
        resolve([NSNumber numberWithInt:idValue]);
    } @catch (NSException *exception) {
        reject(@"JSCollectorElement",@"getID expection",nil);
    }
}

/**
 * 获取采集对象的名称
 * @param collectorId
 * @param promise
 */
RCT_REMAP_METHOD(getName,getName:(NSString*)senderId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        CollectorElement* sender = [JSObjManager getObjWithKey:senderId];
        NSString* name = [sender name];
        resolve(name);
    } @catch (NSException *exception) {
        reject(@"JSCollectorElement",@"getName expection",nil);
    }
}

/**
 * 获取采集对象的备注信息
 * 获取采集对象的名称
 * @param collectorId
 * @param promise
 */
RCT_REMAP_METHOD(getNotes,getNotes:(NSString*)senderId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        CollectorElement* sender = [JSObjManager getObjWithKey:senderId];
        NSString* notes = [sender notes];
        resolve(notes);
    } @catch (NSException *exception) {
        reject(@"JSCollectorElement",@"getNotes expection",nil);
    }
}

/**
 * 获取单击事件监听器
 * @param collectorId
 * @param promise
 */

RCT_REMAP_METHOD(getOnClickListenner,getOnClickListenner:(NSString*)senderId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        CollectorElement* sender = [JSObjManager getObjWithKey:senderId];
         // TODO 获取单击事件监听器
        resolve([NSNumber numberWithBool:YES]);
    } @catch (NSException *exception) {
        reject(@"JSCollectorElement",@"getOnClickListenner expection",nil);
    }
}

/**
 * 获取点串分组信息（仅适用于通过Geomotry构造的动态数据）
 * @param collectorId
 * @param promise
 */
RCT_REMAP_METHOD(getPart,getPart:(NSString*)senderId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        CollectorElement* sender = [JSObjManager getObjWithKey:senderId];
        NSArray* part = [sender getPart];
        NSInteger nsKey = (NSInteger)part;
        [JSObjManager addObj:part];
        resolve(@(nsKey).stringValue);
    } @catch (NSException *exception) {
        reject(@"JSCollectorElement",@"getPart expection",nil);
    }
}


/**
 * 获取采集对象的类型
 * @param collectorId
 * @param promise
 */
RCT_REMAP_METHOD(getType,getType:(NSString*)senderId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        CollectorElement* sender = [JSObjManager getObjWithKey:senderId];
        GPSElementType type = [sender getType];
        resolve([NSNumber numberWithInt:type]);
    } @catch (NSException *exception) {
        reject(@"JSCollectorElement",@"getType expection",nil);
    }
}

/**
 * 获取用户数据
 * @param collectorId
 * @param promise
 */
RCT_REMAP_METHOD(getUserData,getUserData:(NSString*)senderId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        CollectorElement* sender = [JSObjManager getObjWithKey:senderId];
        NSString* data = [sender userData];
        resolve(data);
    } @catch (NSException *exception) {
        reject(@"JSCollectorElement",@"getUserData expection",nil);
    }
}

/**
 * 设置采集对象的名称
 * @param collectorId
 * @param value
 * @param promise
 */
RCT_REMAP_METHOD(setName,setName:(NSString*)senderId name:(NSString*)nameStr resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        CollectorElement* sender = [JSObjManager getObjWithKey:senderId];
        [sender setName:nameStr];
        resolve([NSNumber numberWithBool:YES]);
    } @catch (NSException *exception) {
        reject(@"JSCollectorElement",@"setName expection",nil);
    }
}

/**
 * 设置采集对象的备注信息
 * @param collectorId
 * @param value
 * @param promise
 */
RCT_REMAP_METHOD(setNotes,setNotes:(NSString*)senderId notes:(NSString*)notesStr resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        CollectorElement* sender = [JSObjManager getObjWithKey:senderId];
        [sender setNotes:notesStr];
        resolve([NSNumber numberWithBool:YES]);
    } @catch (NSException *exception) {
        reject(@"JSCollectorElement",@"setNotes expection",nil);
    }
}

/**
 * 设置点击监听器
 * @param collectorId
 * @param value
 * @param promise
 */
RCT_REMAP_METHOD(setOnClickListenner,setOnClickListenner:(NSString*)senderId onClickListenner:(NSString*)listennerId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        CollectorElement* sender = [JSObjManager getObjWithKey:senderId];
        // TODO 设置点击监听器
        resolve([NSNumber numberWithBool:YES]);
    } @catch (NSException *exception) {
        reject(@"JSCollectorElement",@"setOnClickListenner expection",nil);
    }
}

/**
 * 设置用户数据
 * @param collectorId
 * @param value
 * @param promise
 */
RCT_REMAP_METHOD(setUserData,setUserData:(NSString*)senderId data:(NSString*)dataStr resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        CollectorElement* sender = [JSObjManager getObjWithKey:senderId];
        [sender setUserData:dataStr];
        resolve([NSNumber numberWithBool:YES]);
    } @catch (NSException *exception) {
        reject(@"JSCollectorElement",@"setUserData expection",nil);
    }
}

@end
